README.md
# longfi-device-rs
[![Build Status](https://travis-ci.com/helium/longfi-device.svg?token=35YrBmyVB8LNrXzjrRop&branch=master)](https://travis-ci.com/helium/longfi-device-rs)

## Summary

This project creates Rust bindings for the [LongFi Protocol C-library](https://github.com/helium/longfi-device).

## Cloning, building, testing

This project contains the LongFi Protocol C-library as a submodule, so either clone the whole repo and submodules with one command:
	`git clone --recurse-submodules git@github.com:helium/longfi-device-rs.git`

Or, if you've already cloned the top level without submodules, get the submodule:
	`git submodule update --init --recursive`

You can use cargo to build:
	`cargo build [--release]`

The configuration in `.cargo/config` is configured to build only for the `thumbv6m-none-eabi` platform currently.

The code in the example directory is for the [STM32L0 Discovery kit](https://www.st.com/en/evaluation-tools/b-l072z-lrwan1.html), which features the [STM32L072CZ](https://www.st.com/en/microcontrollers-microprocessors/stm32l072cz.html).

To upload the code, start a debug server using either JLink (Note: [you can reprogram the ST-Link](https://www.segger.com/products/debug-probes/j-link/models/other-j-links/st-link-on-board/) on the discovery kit to act like a JLink Server; you will lose the virtual UART over USB provided by the ST-Link):
	`JLinkGDBServer -device STM32L072CZ -speed 4000 -if swd -AutoConnect -1 -port 3333`

or OpenOCD server (Note: if you are using the OpenOCD server, you will want to update `.cargo/config:runner` to use `openocd.gdb` instead of `jlink.gdb`):
	`openocd -f ./openocd.cfg`

run the example:
	`cargo run --example stm32l0x2 [--release]`

Run tests:
    `cargo test --target x86_64-unknown-linux-gnu --tests`


CHANGELOG.md
<!--
M.m.p (YYYY-MM-DD)
==================
Add a summary of this release.

**BREAKING CHANGES**:

* Some change which breaks API or ABI compatiblity with.


Feature enhancements:

* [Link to github PR]():
  A new feature.

Bug fixes:

* [Link to github PR]():
  A bugfix.
-->
0.1.2 (2019-11-26)
==================
* [Workaround for SX126x](https://github.com/helium/longfi-device/pull/26)
Bump longfi-device submodule to allow SX126x to work
* [Include only used radio in binary output](https://github.com/helium/longfi-device-rs/pull/23)
API is changed so that its easier for rustc to only include relevant radio


0.1.1 (2019-11-13)
==================
Fixed bug where fingerprints failed due to memory misalignment with C bindings

0.1.0 (2019-10-28)
==================
Initial release.


src/lib.rs
#![cfg_attr(not(test), no_std)]
use longfi_sys;

pub use longfi_sys::AntPinsMode_t as AntPinsMode;
pub use longfi_sys::BoardBindings_t as BoardBindings;
pub use longfi_sys::ClientEvent_t as ClientEvent;
pub use longfi_sys::LongFiAuthCallbacks as AuthCb;
pub use longfi_sys::LongFiAuthMode_t as AuthMode;
pub use longfi_sys::LongFiConfig_t as Config;
pub use longfi_sys::LongFi_t;
use longfi_sys::Radio_t;
pub use longfi_sys::RfEvent_t as RfEvent;
pub use longfi_sys::RxPacket_t as RxPacket;
pub use longfi_sys::SX126xRadioNew;
pub use longfi_sys::SX1276RadioNew;

static mut SX12XX: Option<Radio> = None;

pub struct LongFi {
    c_handle: LongFi_t,
}

#[derive(Debug)]
pub enum Error {
    NoRadioPointer,
}

unsafe impl Send for LongFi {}

pub struct Radio {
    c_handle: Radio_t,
}

impl Radio {
    pub fn sx1262() -> Radio {
        Radio {
            c_handle: unsafe { SX126xRadioNew() },
        }
    }

    pub fn sx1276() -> Radio {
        Radio {
            c_handle: unsafe { SX1276RadioNew() },
        }
    }
}

impl LongFi {
    pub fn new(
        radio: Radio,
        bindings: &mut BoardBindings,
        config: Config,
        auth_cb_set: &[u8; 16],
    ) -> Result<LongFi, Error> {
        unsafe {
            SX12XX = Some(radio);

            let mut auth_cb = core::mem::zeroed::<AuthCb>();
            *auth_cb.preshared_key.as_mut() = auth_cb_set.as_ptr();

            if let Some(radio) = &mut SX12XX {
                let radio_ptr: *mut Radio_t = &mut radio.c_handle;

                let mut longfi_radio = LongFi {
                    c_handle: longfi_sys::longfi_new_handle(bindings, radio_ptr, config, auth_cb),
                };

                longfi_sys::longfi_init(&mut longfi_radio.c_handle);

                Ok(longfi_radio)
            } else {
                Err(Error::NoRadioPointer)
            }
        }
    }

    pub fn set_buffer(&mut self, buffer: &mut [u8]) {
        unsafe {
            longfi_sys::longfi_set_buf(&mut self.c_handle, buffer.as_mut_ptr(), buffer.len());
        }
    }

    pub fn handle_event(&mut self, event: RfEvent) -> ClientEvent {
        unsafe { longfi_sys::longfi_handle_event(&mut self.c_handle, event) }
    }

    pub fn send(&mut self, buffer: &[u8]) {
        unsafe {
            longfi_sys::longfi_send(&mut self.c_handle, buffer.as_ptr(), buffer.len());
        }
    }

    pub fn send_test(&mut self) {
        unsafe {
            longfi_sys::longfi_rf_test(&mut self.c_handle);
        }
    }

    pub fn get_rx(&mut self) -> RxPacket {
        unsafe { longfi_sys::longfi_get_rx() }
    }
}

extern crate libm;

#[no_mangle]
pub extern "C" fn ceil(expr: f64) -> f64 {
    libm::ceil(expr)
}

#[no_mangle]
pub extern "C" fn round(expr: f64) -> f64 {
    libm::round(expr)
}

#[no_mangle]
pub extern "C" fn floor(expr: f64) -> f64 {
    libm::floor(expr)
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::{LongFi, RfConfig};
    use longfi_sys::{
        GpioIrqHandler, Gpio_t, IrqModes, IrqPriorities, PinConfigs, PinModes, PinNames, PinTypes,
        Spi_t,
    };

    #[no_mangle]
    pub extern "C" fn spi_in_out(s: *mut Spi_t, out_data: u16) -> u16 {
        0
    }
    #[no_mangle]
    pub extern "C" fn gpio_init(
        obj: *mut Gpio_t,
        pin: PinNames,
        mode: PinModes,
        config: PinConfigs,
        pin_type: PinTypes,
        value: u32,
    ) {
    }
    #[no_mangle]
    pub extern "C" fn gpio_write(obj: *mut Gpio_t, value: u32) {}
    #[no_mangle]
    pub extern "C" fn gpio_set_interrupt(
        obj: *mut Gpio_t,
        irq_mode: IrqModes,
        irq_priority: IrqPriorities,
        irq_handler: GpioIrqHandler,
    ) {
    }
    #[no_mangle]
    pub extern "C" fn delay_ms(ms: u32) {}

    static mut RFCONFIG: RfConfig = RfConfig {
        oui: 0x12345678,
        device_id: 0x9abc,
    };
    static mut BINDINGS: BoardBindings = BoardBindings {
        spi_in_out: Some(spi_in_out),
        delay_ms: Some(delay_ms),
        gpio_init: Some(gpio_init),
        gpio_write: Some(gpio_write),
        gpio_set_interrupt: Some(gpio_set_interrupt),
    };
    static mut LONGFI: Option<LongFi> = None;

    #[test]
    fn test_linking() {
        let mut longfi_radio = unsafe { LongFi::new(&mut BINDINGS, &mut RFCONFIG).unwrap() };
        longfi_radio.initialize();

        unsafe { LONGFI = Some(longfi_radio) };
    }

    #[test]
    fn test_sending() {
        let config = RfConfig {
            oui: 0x12345678,
            device_id: 0x9abc,
        };

        let bindings = BoardBindings {
            spi_in_out: Some(spi_in_out),
            delay_ms: Some(delay_ms),
            gpio_init: Some(gpio_init),
            gpio_write: Some(gpio_write),
            gpio_set_interrupt: Some(gpio_set_interrupt),
        };

        let packet: [u8; 5] = [0xDE, 0xAD, 0xBE, 0xEF, 0];
        unsafe {
            if let Some(longfi) = &mut LONGFI {
                longfi.send(&packet);
            }
        };
    }
}


longfi-sys/build.rs
#[cfg(workaround_build)]
fn main() {
    use std::env;
    use std::path::PathBuf;
    use std::process::Command;
    use cmake;
    use cmake::Config;

    let dst = Config::new("longfi-device")
                 .define("BUILD_TESTING", "OFF")
                 .define("CMAKE_C_COMPILER_WORKS", "1")
                 .define("CMAKE_CXX_COMPILER_WORKS", "1")
                 .pic(false)
                 .build();

    println!("cargo:rustc-link-search=native={}/lib", dst.display());
    println!("cargo:rustc-link-lib=static=blake2");
    println!("cargo:rustc-link-lib=static=cursor");
    println!("cargo:rustc-link-lib=static=cursor_varint");
    println!("cargo:rustc-link-lib=static=golay");
    println!("cargo:rustc-link-lib=static=lfc");
    println!("cargo:rustc-link-lib=static=longfi");
    println!("cargo:rustc-link-lib=static=sx12xx");

   // make the bindings
   let bindings = bindgen::Builder::default()
       .raw_line("use cty;")
       .use_core()
       .ctypes_prefix("cty")
       .detect_include_paths(true)
       .header("longfi-device/board.h")
       .header("longfi-device/longfi.h")
       .header("longfi-device/radio/radio.h")
       .header("longfi-device/radio/sx1276/sx1276.h")
       .header("longfi-device/radio/sx126x/sx126x.h")
       .clang_arg(format!("-I{}/include",dst.display()))
       .whitelist_var("XTAL_FREQ")
       .whitelist_var("FREQ_STEP")
       .whitelist_var("RX_BUFFER_SIZE")
       .whitelist_type("RfEvent_t")
       .whitelist_type("RadioState_t")
       .whitelist_type("RadioModems_t")
       .whitelist_type("ClientEvent_t")
       .whitelist_type("QualityOfService_t")
       .whitelist_type("RfConfig_t")
       .whitelist_type("RxPacket_t")
       .whitelist_type("LF_Gpio_t")
       .whitelist_type("LF_Spi_t")
       .whitelist_type("AntPinsMode_t")
       .whitelist_type("LongFiAuthMode_t")
       .whitelist_type("LongFiAuthCallbacks")
       .whitelist_function("longfi_init")
       .whitelist_function("longfi_new_handle")
       .whitelist_function("longfi_handle_event")
       .whitelist_function("longfi_send")
       .whitelist_function("longfi_get_rx")
       .whitelist_function("longfi_set_buf")
       .whitelist_function("longfi_rf_test")
       .whitelist_function("longfi_get_random")
       .whitelist_function("longfi_enable_tcxo")
       .whitelist_function("board_set_bindings")
       .whitelist_function("memcpy1")
       .whitelist_function("SX1276RadioNew")
       .whitelist_function("SX126xRadioNew")
       .whitelist_function("SX126xReadRegister")
       .trust_clang_mangling(false)
       .rustfmt_bindings(true)
       .rustified_enum("ClientEvent_t")
       .rustified_enum("RfEvent_t")
       .rustified_enum("QualityOfService_t")
       .rustified_enum("AntPinsMode_t")
       .rustified_enum("LongFiAuthMode_t")
       .derive_copy(false)
       .derive_debug(false)
       .layout_tests(false)
       .generate()
       .expect("Failed to generate sx1276 bindings!");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}

#[cfg(not(workaround_build))]
fn main() {
   cargo_5730::run_build_script();
}


examples/stm32l0x2/longfi_linking.rs
use core::ffi;
use embedded_hal::digital::v2::OutputPin;
use embedded_hal::spi::FullDuplex;
use longfi_sys::{
    GpioIrqHandler, Gpio_t, IrqModes, IrqPriorities, PinConfigs, PinModes, PinNames, PinTypes,
    Spi_t,
};
use nb::block;
use stm32l0xx_hal as hal;
use stm32l0xx_hal::gpio::gpioa::*;
use stm32l0xx_hal::gpio::{Floating, Input, Output, PushPull};
use stm32l0xx_hal::pac::SPI1;

#[no_mangle]
pub extern "C" fn spi_in_out(s: *mut Spi_t, out_data: u16) -> u16 {
    let spi: &mut hal::spi::Spi<
        SPI1,
        (
            PA3<Input<Floating>>,
            PA6<Input<Floating>>,
            PA7<Input<Floating>>,
        ),
    > = unsafe {
        &mut *((*s).Spi.Instance
            as *mut hal::spi::Spi<
                SPI1,
                (
                    PA3<Input<Floating>>,
                    PA6<Input<Floating>>,
                    PA7<Input<Floating>>,
                ),
            >)
    };

    spi.send(out_data as u8).unwrap();
    let in_data = block!(spi.read()).unwrap();

    in_data as u16
}

#[no_mangle]
pub extern "C" fn gpio_init(
    obj: *mut Gpio_t,
    pin: PinNames,
    mode: PinModes,
    config: PinConfigs,
    pin_type: PinTypes,
    value: u32,
) {
}
#[no_mangle]
pub extern "C" fn gpio_write(obj: *mut Gpio_t, value: u32) {
    let gpio: &mut stm32l0xx_hal::gpio::gpioa::PA15<Output<PushPull>> =
        unsafe { &mut *(obj as *mut stm32l0xx_hal::gpio::gpioa::PA15<Output<PushPull>>) };

    if value == 0 {
        gpio.set_low().unwrap();
    } else {
        gpio.set_high().unwrap();
    }
}

#[no_mangle]
pub extern "C" fn gpio_set_interrupt(
    obj: *mut Gpio_t,
    irq_mode: IrqModes,
    irq_priority: IrqPriorities,
    irq_handler: GpioIrqHandler,
) {
}

#[no_mangle]
pub extern "C" fn delay_ms(ms: u32) {}


examples/stm32l0x2/longfi_bindings.rs
use hal::device;
use hal::exti;
use hal::gpio::*;
use hal::pac;
use hal::prelude::*;
use hal::rcc::Rcc;
use hal::rng;
use hal::spi;
use longfi_device::{AntPinsMode, BoardBindings};
use nb::block;
use stm32l0xx_hal as hal;

#[allow(dead_code)]
pub struct LongFiBindings {
    pub bindings: BoardBindings,
}

type Uninitialized = Input<Floating>;

pub type RadioIRQ = gpiob::PB4<Input<PullUp>>;

pub fn initialize_irq(
    pin: gpiob::PB4<Uninitialized>,
    syscfg: &mut hal::syscfg::SYSCFG,
    exti: &mut pac::EXTI,
) -> gpiob::PB4<Input<PullUp>> {
    let dio0 = pin.into_pull_up_input();

    exti.listen(
        syscfg,
        dio0.port(),
        dio0.pin_number(),
        exti::TriggerEdge::Rising,
    );

    dio0
}

pub type TcxoEn = gpioa::PA8<Output<PushPull>>;

impl LongFiBindings {
    pub fn new(
        spi_peripheral: device::SPI1,
        rcc: &mut Rcc,
        rng: rng::Rng,
        spi_sck: gpiob::PB3<Uninitialized>,
        spi_miso: gpioa::PA6<Uninitialized>,
        spi_mosi: gpioa::PA7<Uninitialized>,
        spi_nss_pin: gpioa::PA15<Uninitialized>,
        reset: gpioc::PC0<Uninitialized>,
        rx: gpioa::PA1<Uninitialized>,
        tx_rfo: gpioc::PC2<Uninitialized>,
        tx_boost: gpioc::PC1<Uninitialized>,
    ) -> LongFiBindings {
        // store all of the necessary pins and peripherals into statics
        // this is necessary as the extern C functions need access
        // this is safe, thanks to ownership and because these statics are private
        unsafe {
            SPI = Some(spi_peripheral.spi(
                (spi_sck, spi_miso, spi_mosi),
                spi::MODE_0,
                1_000_000.hz(),
                rcc,
            ));
            SPI_NSS = Some(spi_nss_pin.into_push_pull_output());
            RESET = Some(reset.into_push_pull_output());
            ANT_SW = Some(AntennaSwitches::new(
                rx.into_push_pull_output(),
                tx_rfo.into_push_pull_output(),
                tx_boost.into_push_pull_output(),
            ));
            RNG = Some(rng);
        };

        LongFiBindings {
            bindings: BoardBindings {
                reset: Some(radio_reset),
                spi_in_out: Some(spi_in_out),
                spi_nss: Some(spi_nss),
                delay_ms: Some(delay_ms),
                get_random_bits: Some(get_random_bits),
                set_antenna_pins: Some(set_antenna_pins),
                set_board_tcxo: None,
                busy_pin_status: None,
                reduce_power: None,
            },
        }
    }
}

static mut EN_TCXO: Option<TcxoEn> = None;

#[no_mangle]
pub extern "C" fn set_tcxo(value: bool) -> u8 {
    unsafe {
        if let Some(pin) = &mut EN_TCXO {
            if value {
                pin.set_high().unwrap();
            } else {
                pin.set_low().unwrap();
            }
        }
    }
    6
}

type SpiPort = hal::spi::Spi<
    hal::pac::SPI1,
    (
        hal::gpio::gpiob::PB3<hal::gpio::Input<hal::gpio::Floating>>,
        hal::gpio::gpioa::PA6<hal::gpio::Input<hal::gpio::Floating>>,
        hal::gpio::gpioa::PA7<hal::gpio::Input<hal::gpio::Floating>>,
    ),
>;
static mut SPI: Option<SpiPort> = None;
#[no_mangle]
extern "C" fn spi_in_out(out_data: u8) -> u8 {
    unsafe {
        if let Some(spi) = &mut SPI {
            spi.send(out_data).unwrap();
            let in_data = block!(spi.read()).unwrap();
            in_data
        } else {
            0
        }
    }
}

static mut SPI_NSS: Option<gpioa::PA15<Output<PushPull>>> = None;
#[no_mangle]
extern "C" fn spi_nss(value: bool) {
    unsafe {
        if let Some(pin) = &mut SPI_NSS {
            if value {
                pin.set_high().unwrap();
            } else {
                pin.set_low().unwrap();
            }
        }
    }
}

static mut RESET: Option<gpioc::PC0<Output<PushPull>>> = None;
#[no_mangle]
extern "C" fn radio_reset(value: bool) {
    unsafe {
        if let Some(pin) = &mut RESET {
            if value {
                pin.set_low().unwrap();
            } else {
                pin.set_high().unwrap();
            }
        }
    }
}

#[no_mangle]
extern "C" fn delay_ms(ms: u32) {
    cortex_m::asm::delay(ms);
}

static mut RNG: Option<rng::Rng> = None;
extern "C" fn get_random_bits(_bits: u8) -> u32 {
    unsafe {
        if let Some(rng) = &mut RNG {
            // enable starts the ADC conversions that generate the random number
            rng.enable();
            // wait until the flag flips; interrupt driven is possible but no implemented
            rng.wait();
            // reading the result clears the ready flag
            let val = rng.take_result();
            // can save some power by disabling until next random number needed
            rng.disable();
            val
        } else {
            panic!("No Rng exists!");
        }
    }
}

pub struct AntennaSwitches<Rx, TxRfo, TxBoost> {
    rx: Rx,
    tx_rfo: TxRfo,
    tx_boost: TxBoost,
}
#[warn(unused_must_use)]
impl<Rx, TxRfo, TxBoost> AntennaSwitches<Rx, TxRfo, TxBoost>
where
    Rx: embedded_hal::digital::v2::OutputPin,
    TxRfo: embedded_hal::digital::v2::OutputPin,
    TxBoost: embedded_hal::digital::v2::OutputPin,
{
    pub fn new(rx: Rx, tx_rfo: TxRfo, tx_boost: TxBoost) -> AntennaSwitches<Rx, TxRfo, TxBoost> {
        AntennaSwitches {
            rx,
            tx_rfo,
            tx_boost,
        }
    }

    pub fn set_sleep(&mut self) {
        self.rx.set_low().unwrap_or(());
        self.tx_rfo.set_low().unwrap_or(());
        self.tx_boost.set_low().unwrap_or(());
    }

    pub fn set_tx(&mut self) {
        self.rx.set_low().unwrap_or(());
        self.tx_rfo.set_low().unwrap_or(());
        self.tx_boost.set_high().unwrap_or(());
    }

    pub fn set_rx(&mut self) {
        self.rx.set_high().unwrap_or(());
        self.tx_rfo.set_low().unwrap_or(());
        self.tx_boost.set_low().unwrap_or(());
    }
}

type AntSw = AntennaSwitches<
    stm32l0xx_hal::gpio::gpioa::PA1<stm32l0xx_hal::gpio::Output<stm32l0xx_hal::gpio::PushPull>>,
    stm32l0xx_hal::gpio::gpioc::PC2<stm32l0xx_hal::gpio::Output<stm32l0xx_hal::gpio::PushPull>>,
    stm32l0xx_hal::gpio::gpioc::PC1<stm32l0xx_hal::gpio::Output<stm32l0xx_hal::gpio::PushPull>>,
>;

static mut ANT_SW: Option<AntSw> = None;

pub extern "C" fn set_antenna_pins(mode: AntPinsMode, _power: u8) {
    unsafe {
        if let Some(ant_sw) = &mut ANT_SW {
            match mode {
                AntPinsMode::AntModeTx => {
                    ant_sw.set_tx();
                }
                AntPinsMode::AntModeRx => {
                    ant_sw.set_rx();
                }
                AntPinsMode::AntModeSleep => {
                    ant_sw.set_sleep();
                }
                _ => (),
            }
        }
    }
}


examples/stm32l0x2/main.rs
#![cfg_attr(not(test), no_std)]
#![no_main]

// To use example, press any key in serial terminal
// Packet will send and "Transmit Done!" will print when radio is done sending packet

extern crate nb;
extern crate panic_halt;

use core::fmt::Write;
use hal::serial::USART2 as DebugUsart;
use hal::{pac, prelude::*, rcc, rng::Rng, serial, syscfg};
use longfi_device;
use longfi_device::{ClientEvent, Config, LongFi, Radio, RfEvent};
use stm32l0xx_hal as hal;

mod longfi_bindings;
pub use longfi_bindings::initialize_irq as initialize_radio_irq;
pub use longfi_bindings::LongFiBindings;
pub use longfi_bindings::RadioIRQ;
pub use longfi_bindings::TcxoEn;

const OUI: u32 = 1;
const DEVICE_ID: u16 = 5;
const PRESHARED_KEY: [u8; 16] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];

#[rtfm::app(device = stm32l0xx_hal::pac)]
const APP: () = {
    static mut INT: pac::EXTI = ();

    static mut RADIO_IRQ: RadioIRQ = ();
    static mut DEBUG_UART: serial::Tx<DebugUsart> = ();
    static mut UART_RX: serial::Rx<DebugUsart> = ();
    static mut BUFFER: [u8; 512] = [0; 512];
    static mut COUNT: u8 = 0;
    static mut LONGFI: LongFi = ();

    #[init(spawn = [send_ping], resources = [BUFFER])]
    fn init() -> init::LateResources {
        static mut BINDINGS: Option<LongFiBindings> = None;

        let mut rcc = device.RCC.freeze(rcc::Config::hsi16());
        let mut syscfg = syscfg::SYSCFG::new(device.SYSCFG, &mut rcc);

        let gpioa = device.GPIOA.split(&mut rcc);
        let gpiob = device.GPIOB.split(&mut rcc);
        let gpioc = device.GPIOC.split(&mut rcc);

        let (tx_pin, rx_pin, serial_peripheral) = (gpioa.pa2, gpioa.pa3, device.USART2);

        let mut serial = serial_peripheral
            .usart((tx_pin, rx_pin), serial::Config::default(), &mut rcc)
            .unwrap();

        // listen for incoming bytes which will trigger transmits
        serial.listen(serial::Event::Rxne);
        let (mut tx, rx) = serial.split();

        write!(tx, "LongFi Device Test\r\n").unwrap();

        let mut exti = device.EXTI;
        // constructor initializes 48 MHz clock that RNG requires
        // Initialize 48 MHz clock and RNG
        let hsi48 = rcc.enable_hsi48(&mut syscfg, device.CRS);
        let rng = Rng::new(device.RNG, &mut rcc, hsi48);
        let radio_irq = initialize_radio_irq(gpiob.pb4, &mut syscfg, &mut exti);

        *BINDINGS = Some(LongFiBindings::new(
            device.SPI1,
            &mut rcc,
            rng,
            gpiob.pb3,
            gpioa.pa6,
            gpioa.pa7,
            gpioa.pa15,
            gpioc.pc0,
            gpioa.pa1,
            gpioc.pc2,
            gpioc.pc1,
        ));

        let rf_config = Config {
            oui: OUI,
            device_id: DEVICE_ID,
            auth_mode: longfi_device::AuthMode::PresharedKey128,
        };

        let mut longfi_radio;
        if let Some(bindings) = BINDINGS {
            longfi_radio = LongFi::new(
                Radio::sx1276(),
                &mut bindings.bindings,
                rf_config,
                &PRESHARED_KEY,
            )
            .unwrap()
        } else {
            panic!("No bindings exist");
        }

        longfi_radio.set_buffer(resources.BUFFER);

        write!(tx, "Going to main loop\r\n").unwrap();

        // Return the initialised resources.
        init::LateResources {
            INT: exti,
            RADIO_IRQ: radio_irq,
            DEBUG_UART: tx,
            UART_RX: rx,
            LONGFI: longfi_radio,
        }
    }

    #[task(capacity = 4, priority = 2, resources = [DEBUG_UART, BUFFER, LONGFI])]
    fn radio_event(event: RfEvent) {
        let mut longfi_radio = resources.LONGFI;
        let client_event = longfi_radio.handle_event(event);

        match client_event {
            ClientEvent::ClientEvent_TxDone => {
                write!(resources.DEBUG_UART, "Transmit Done!\r\n").unwrap();
            }
            ClientEvent::ClientEvent_Rx => {
                // get receive buffer
                let rx_packet = longfi_radio.get_rx();
                write!(resources.DEBUG_UART, "Received packet\r\n").unwrap();
                write!(resources.DEBUG_UART, "  Length =  {}\r\n", rx_packet.len).unwrap();
                write!(resources.DEBUG_UART, "  Rssi   = {}\r\n", rx_packet.rssi).unwrap();
                write!(resources.DEBUG_UART, "  Snr    =  {}\r\n", rx_packet.snr).unwrap();
                unsafe {
                    for i in 0..rx_packet.len {
                        write!(
                            resources.DEBUG_UART,
                            "{:X} ",
                            *rx_packet.buf.offset(i as isize)
                        )
                        .unwrap();
                    }
                    write!(resources.DEBUG_UART, "\r\n").unwrap();
                }
                // give buffer back to library
                longfi_radio.set_buffer(resources.BUFFER);
            }
            ClientEvent::ClientEvent_None => {}
        }
    }

    #[task(capacity = 4, priority = 2, resources = [DEBUG_UART, COUNT, LONGFI])]
    fn send_ping() {
        write!(resources.DEBUG_UART, "Sending Ping\r\n").unwrap();
        let packet: [u8; 72] = [
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            *resources.COUNT,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xa1,
            0xa2,
            0xa3,
            0xa4,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            *resources.COUNT,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xa1,
            0xa2,
            0xa3,
            0xa4,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xDE,
            0xAD,
            0xBE,
            0xEF,
            0xa1,
            0xa2,
            0xa3,
            0xa4,
            0xBE,
            0xEF,
            0xa1,
            0xa2,
            0xa3,
            0xa4,
        ];
        *resources.COUNT += 1;
        resources.LONGFI.send(&packet);
    }

    #[interrupt(priority=1, resources = [UART_RX], spawn = [send_ping])]
    fn USART2() {
        let rx = resources.UART_RX;
        rx.read().unwrap();
        spawn.send_ping().unwrap();
    }

    #[interrupt(priority = 1, resources = [RADIO_IRQ, INT], spawn = [radio_event])]
    fn EXTI4_15() {
        resources.INT.clear_irq(resources.RADIO_IRQ.pin_number());
        spawn.radio_event(RfEvent::DIO0).unwrap();
    }

    // Interrupt handlers used to dispatch software tasks
    extern "C" {
        fn USART4_USART5();
    }
};
